feat(forms): add markAllAsDirty to AbstractControl

aberonni opened the pull request:
Adds the `markAllAsDirty` method to the `AbstractControl` class. This method will mark the control and all its descendants as dirty.

I pretty much just duplicated the behaviour and tests of `markAllAsTouched`.

Fixes #55990

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #55990 


## What is the new behavior?

Adding `markAllAsDirty` to `AbstractControl`

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


## Other information

N/A

aberonni committed:
feat(forms): add markAllAsDirty to AbstractControl

Adds the `markAllAsDirty` method to the `AbstractControl` class. This method will mark the control and all its
descendants as dirty.

I pretty much just duplicated the behaviour and tests of `markAllAsTouched`.

Fixes #55990

aberonni commented:
@AndrewKushnir just checking if this requires any action from me?

michael-small commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -994,6 +994,22 @@ export abstract class AbstractControl<TValue = any, TRawValue extends TValue = T
+   * @param opts Configuration options that determine how the control propagates changes
+   * and emits events after marking is applied.
+   * * `emitEvent`: When true or not supplied (the default), the `events`
+   * observable emits a `PristineChangeEvent` with the `pristing` property being `false`.
```
```suggestion
   * observable emits a `PristineChangeEvent` with the `pristine` property being `false`.
```

JeanMeche commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -198,6 +198,59 @@ import {asyncValidator} from './util';
+
+        expect(innerFormArrayGroupControl1.dirty).toBe(false);
+
+        formArray.markAllAsDirty();
```
Can we add tests that also cover  `emitEvent` ?  (and checking the event are emitted)

aberonni replied:
I've added a basic test which I grabbed from elsewhere for inspiration. I also did the same for `markAllAsTouched` which is what I got inspiration from for the other tests.

michael-small commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -198,6 +198,59 @@ import {asyncValidator} from './util';
+
+        expect(group1.dirty).toBe(false);
+
+        const group1Control1 = group1.get('c1') as FormControl;
```
This question may be bigger than this PR's scope:

I see that the old fashioned `.get('...') as FormXXX` conventions come from existing tests. But I wonder: is there any reason for this now other than consistency? I would think they could be properly typed now by just accessing the now default typing. Either just for these new tests now, or perhaps take a more formal sweep through form tests in a separate PR.

aberonni replied:
Tbh I just followed what was in the tests for `markAllAsTouched`. This is my first contribution to angular so I'm not sure how else to do things, but happy to adjust if this is deemed within the scope of my contribution.

michael-small reviewed (approved):
LGTM

kirjs commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -141,6 +221,12 @@ import {StatusChangeEvent} from '../src/model/abstract_model';
         expect(arrayFirstChildGroupFirstChildCtrl.touched).toBe(true);
       });
+
+      it('should not emit events when `FormGroup.markAllAsTouched` is called with `emitEvent: false`', () => {
```
Thanks for adding this test.
Consider also adding a test that verifies that event is emitted for true and not having a value.

kirjs commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -994,6 +994,22 @@ export abstract class AbstractControl<TValue = any, TRawValue extends TValue = T
+   * When false, no events are emitted.
+   */
+  markAllAsDirty(opts: {emitEvent?: boolean} = {}): void {
+    this.markAsDirty({onlySelf: true, emitEvent: opts.emitEvent, sourceControl: this});
```
I'm wondering why onlySelf is not being passed as a prop to cover the case where we want to mark descendants but not ascendance?

I assume this is just something you copypasted from `markAllAsTouched`, but curious if you have any context for this.

kirjs commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -81,9 +81,86 @@ import {StatusChangeEvent} from '../src/model/abstract_model';
+        expect(arrayFirstChildGroupFirstChildCtrl.dirty).toBe(true);
+      });
+
+      it('should not emit events when `FormGroup.markAllAsDirty` is called with `emitEvent: false`', () => {
```
Please add a test that verifies that event is emitted for true and not having a value.

kirjs commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -994,6 +994,22 @@ export abstract class AbstractControl<TValue = any, TRawValue extends TValue = T
+   *
+   * @param opts Configuration options that determine how the control propagates changes
+   * and emits events after marking is applied.
+   * * `emitEvent`: When true or not supplied (the default), the `events`
```
<img width="775" alt="image" src="https://github.com/user-attachments/assets/b35a3ac7-a497-4c02-8506-a95df6a90881" />
Looking at the code, I'm not convinced that true and missing value would do the same thing, but a test would help clarify that

JeanMeche replied:
`undefined` is `!== false`. 
I think we're good here !

aberonni commented:
@kirjs thanks for the review. I'm not sure how to proceed on this, given that the PR seems to have the necessary approvals, but there are also change requests. I also want to point out that the requested changes seem to be beyond the scope of my addition - given that they are addressing feedback on tests that were already there. If it helps get this merged, I can put more time and effort in, but I doubt I am the best suited to put that effort in (I'm a first-time contributor, and just trying to add a very small change).

michael-small commented:
I'm just a random community guy who keeps up with forms PRs. I think these changes are legit for what it's worth but I'm not blocking anything.  My comments were just thinking out loud to real maintainers about the existing tests beyond the scope of your PR. 

Looks good to me for what it's worth

JeanMeche commented:
Pullapprove is reporting that we miss 2 public-api scope approvals.

thePunderWoman reviewed (approved):
reviewed-for: public-api

atscott reviewed (approved):
reviewed-for: public-api

thePunderWoman commented:
This PR was merged into the repository by commit a07ee60989441c38e6539fd25cad5166622e9f9e.

The changes were merged into the following branches: main
